### PR TITLE
backend: Remove logger nil checks

### LINKED ISSUE #19015: Possible missing nil checks
### Bug report criteria

- [X] This bug report is not security related, security issues should be disclosed privately via [etcd maintainers](mailto:etcd-maintainers@googlegroups.com).
- [X] This is not a support request or question, support requests or questions should be raised in the etcd [discussion forums](https://github.com/etcd-io/etcd/discussions).
- [X] You have read the etcd [bug reporting guidelines](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/reporting_bugs.md).
- [X] Existing open issues along with etcd [frequently asked questions](https://etcd.io/docs/latest/faq) have been checked and this is not a duplicate.

### What happened?

After having been compared to a nil value at backend.go:515, pointer 'b.lg' is passed as implicit 'this' parameter in call to function 'Error' at backend.go:545 or pointer 'b.lg' is passed as implicit 'this' parameter in call to function 'Fatal' at backend.go:532. 
```
if b.lg != nil {
		b.lg.Info(
			"defragmenting",
			zap.String("path", dbp),
			zap.Int64("current-db-size-bytes", size1),
			zap.String("current-db-size", humanize.Bytes(uint64(size1))),
			zap.Int64("current-db-size-in-use-bytes", sizeInUse1),
			zap.String("current-db-size-in-use", humanize.Bytes(uint64(sizeInUse1))),
		)
	}

	defer func() {
		// NOTE: We should exit as soon as possible because that tx
		// might be closed. The inflight request might use invalid
		// tx and then panic as well. The real panic reason might be
		// shadowed by new panic. So, we should fatal here with lock.
		if rerr := recover(); rerr != nil {
			b.lg.Fatal("unexpected panic during defrag", zap.Any("panic", rerr))
		}
	}()

	// Commit/stop and then reset current transactions (including the readTx)
	b.batchTx.unsafeCommit(true)
	b.batchTx.tx = nil

	// gofail: var defragBeforeCopy struct{}
	err = defragdb(b.db, tmpdb, defragLimit)
	if err != nil {
		tmpdb.Close()
		if rmErr := os.RemoveAll(tmpdb.Path()); rmErr != nil {
			b.lg.Error("failed to remove db.tmp after defragmentation completed", zap.Error(rmErr))
		}

		// restore the bbolt transactions if defragmentation fails
		b.batchTx.tx = b.unsafeBegin(true)
		b.readTx.tx = b.unsafeBegin(false)

		return err
	}
```
Please clarify whether verification is required in this case? Do you need to make any changes/corrections to the code?

### What did you expect to happen?

It might be worth adding additional checks for b.lg, if required.

### How can we reproduce it (as minimally and precisely as possible)?

Checking code

### Anything else we need to know?

_No response_

### Etcd version (please run commands below)

Latest

### Etcd configuration (command line flags or environment variables)

<details>

</details>

### Etcd debug information (please run commands below, feel free to obfuscate the IP address or FQDN in the output)

<details>

</details>

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by srujangit123**
To remove the ugly nil checks by instantiating the logger with a no-op logger if it is nil.

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

Fixes #19015

### CODE REVIEW

**Comment by ahrtr** [comment]
Please add an assert at the first line of `(b *backend) defrag()`, something like below,

https://github.com/etcd-io/etcd/blob/b4450d510ddf0f011c874c28e95175b4bac7f6c8/pkg/osutil/interrupt_unix.go#L51

**Comment by ahrtr** [comment]
cc @fuweid  @ivanvc @jmhbnz @serathius

File: server/storage/backend/backend.go [hunk 1]
```diff
@@ -199,6 +200,10 @@ func newBackend(bcfg BackendConfig) *backend {
 		bcfg.Logger.Panic("failed to open database", zap.String("path", bcfg.Path), zap.Error(err))
 	}
 
+	if bcfg.Logger == nil {
```
Comment (fuweid) [inline server/storage/backend/backend.go:203]: I don't think it works. Both the `newBoltLoggerZap` call and Line 200 `panic` already use `bcfg.Logger`.
That assignment should be placed at the beginning of the `newBackend`

Comment (ahrtr) [reply server/storage/backend/backend.go:203]: Good catch!

**Review by ahrtr — CHANGES_REQUESTED** [review]
https://github.com/etcd-io/etcd/pull/19077/files#r1889186695

**Review by ahrtr — APPROVED** [review]

**Review by ivanvc — APPROVED** [review]
LGTM. Thanks, @srujangit123.

**Review by serathius — APPROVED** [review]

**Pull request merged**